Stream output files instead of writing them in-memory first

Mikolaj-A-Kowalski opened the pull request:
This resolves Issue #38.  However, I cannot help but to think it is a bit filthy... that being said I am not sure how exactly to make it nicer. 

The problem with streaming to the output file directly is that it is convenient to have a 'backtrace' capability, when writing the printers for different output formats. This allows, for example, to remove trailing commas with relative ease (crucial for JSON). 

To 'have a cake and eat it' this implements a `delayedStream` which works a bit like a buffered output, but when the buffer is flushed it retains `MIN_SIZE=16` last characters [as opposed to printing everything to the disk]. 

What I don't like about the changes is the way the file ownership is handled. It is made to be a responsibility of the `outputFile` class, which opens a file in `init`and closes it in `finalisation` (which is also the destructor). However, the printing is implemented by the `asciiOutput_inter` which owns an instance of `delayedStream`. The appropriate output unit is set by a subroutine.  

The other thing I fear may be confusing is that if one does not supply `filename` to `outputFile` no file will be produced (this is required for unit testing). 

In short this PR may still require some iterations [in addition to squashing the fixups!]

valeriaRaffuzzi commented on UserInterface/fileOutput/asciiJSON_class.f90:
```diff
@@ -71,51 +70,43 @@ subroutine init(self)
 
-    form = '(A' // numToChar(self % output % length()) // ')'
+  !!
+  !! Return approperiate extension for the file
```
```suggestion
  !! Return appropriate extension for the file
```

valeriaRaffuzzi commented on UserInterface/fileOutput/asciiMATLAB_class.f90:
```diff
@@ -90,33 +87,29 @@ subroutine init(self)
 
   !!
-  !! Print the output to the given unit
+  !! Return approperiate extension for the file
```
```suggestion
  !! Return appropriate extension for the file
```

valeriaRaffuzzi commented on UserInterface/fileOutput/dummyPrinter_class.f90:
```diff
@@ -23,6 +23,7 @@ module dummyPrinter_class
     procedure :: init
+    procedure :: endFile
     procedure :: extension
     procedure :: writeToFile
```
I thought writeToFile was gone in the asciiOutput class?

Mikolaj-A-Kowalski replied:
It is... I think this is an oversight that this got left in.

Mikolaj-A-Kowalski replied:
Should be removed now.

valeriaRaffuzzi commented on UserInterface/fileOutput/outputFile_class.f90:
```diff
@@ -210,17 +211,54 @@ module outputFile_class
+      ! It causes buggy warnings to appear in compilation
+      allocate(self % outputFileName, source = trim(filename) // "." // self % output % extension())
+
+      open(newunit = self % outputUnit, &
```
It looks like 'error' is gone from this list.

valeriaRaffuzzi replied:
same with errorMsg

Mikolaj-A-Kowalski replied:
:fearful: Ups... not great thing to forget

Mikolaj-A-Kowalski replied:
Should be fixed.

valeriaRaffuzzi commented on UserInterface/fileOutput/outputFile_class.f90:
```diff
@@ -233,58 +271,32 @@ subroutine init(self, type, fatalErrors)
-    ! Close file
-    close(unitNum)
+  !! NOTE:
+  !!   This subroutine WILL NOT be called if the `end program` statement is
```
This comment is not clear.

Mikolaj-A-Kowalski replied:
```Fortran
  !!  NOTE:
  !!    Being marked `final`(a deconstructor)  this subroutine will be normally implicitly called by
  !!    by the compiler whenever a variable of `type(outputFile)` goes out of scope (e.g. when program 
  !!    execution reaches `end subroutine`, `end function` or `end block` statement. However, the rule in 
  !!    Fortran is that the decostructors are NOT called at the end of the program `end program`. In that case 
  !!    if one uses the output file as a module variable or defines it inside the program block, the `finalisation` 
  !!    must be called explicitly to ensure or data from the buffer is written to the disk 
  !!     
```
Does this work better?

Mikolaj-A-Kowalski replied:
Comment has been changed.

valeriaRaffuzzi reviewed (changes requested):
See comments.

Mikolaj-A-Kowalski commented:
So I think I have addressed all the issues now... 
If it looks ok I will rebase on the current main (and squash all the fixups!) and then it should be ready to be merged.

valeriaRaffuzzi commented on docs/Output.rst:
```diff
@@ -76,9 +77,10 @@ Writing to an output file in SCONE is done through a series of calls to appropri
-  !! Extension will be determined by the format printer
-  call out % writeToFile("./path/to/output/without_any_extension")
+  !! We need to make sure that output file is properly finalised. We can skip
+  !! this line if we we are not in a `program` block e.g. in `subroutine` or `function`
```
```suggestion
  !! this line if we are not in a `program` block e.g. in `subroutine` or `function`
```
Typo!

valeriaRaffuzzi reviewed (approved):
Very nice, I think this all looks great and it's a much better alternative to write the output file than keep it all in memory! 

I think there's just a small leftover typo that I spotted in the docs (soon SCONE will be typo free!!!) but then feel free to rebase and merge
